campaignion_tracking/package/src/ga4.js

```diff
@@ -0,0 +1,426 @@
+/**
+ * Machine readable name.
+ *
+ * Use for referencing tracker specifics like storage keys, channel names, ...
+ */
+export const name = 'ga4'
+
+/**
+ * Mapping of codes to eventNames.
+ *
+ * This also is a whitelist of which events should *also* be recognized as
+ * codes.
+ */
+export const codes = {
+  ds: 'donationSuccess',
+  s: 'submission',
+}
+
+/**
+ * Compare two donation product objects.
+ *
+ * Keys considered: 'name', 'price', 'id', 'quantity'
+ *
+ * Comparing with JSON.stringify will not work robustly as the order is not
+ * guaranteed.
+ *
+ * @param {object} product1 first product
+ * @param {object} product2 second product
+ * @returns {Boolean}
+ */
+export const productIsEqual = (product1, product2) => {
+  for (const key of ['name', 'price', 'id', 'quantity']) {
+    // If at least one product has the property and their values differ
+    // they are NOT equal. They are otherwise.
+    if ((Object.prototype.hasOwnProperty.call(product1, key) || Object.prototype.hasOwnProperty.call(product2, key)) &&
+      product1[key] !== product2[key]
+    ) {
+      return false
+    }
+  }
+  return true
+}
+
+/**
+ * Tracker for Google Tag Manager.
+ *
+ * Implements behaviour to dispatch events to Google Analytics.
+ */
+export class GA4Tracker {
+  /**
+   * Constructor.
+   *
+   * @param {TrackerManager} tracker the shared tracker manager object
+   * @param {object} dataLayer the datalayer to use (default `window.dataLayer`)
+   * @param {Boolean} debug set to true for debugging
+   */
+  constructor (tracker, dataLayer = window.dataLayer, debug = false) {
+    this.debug = debug
+    this.dataLayer = dataLayer
+    this.tracker = tracker
+
+    // load from session, set defaults
+    const defaultContext = {
+      node: {},
+      donation: { currencyCode: null, product: null },
+      webform: { sid: null }
+    }
+    this._context = this.loadFromStorage('context') || defaultContext
+
+    this.printDebug('init')
+
+    if (typeof this.tracker === 'undefined') {
+      this.printDebug('No TrackerManager found. Doing nothing.')
+      return
+    }
+
+    if (typeof this.dataLayer === 'undefined') {
+      this.printDebug('No datalayer found. Doing nothing.')
```

Comment:
Is there a return `return` missing after this statement?

Reply:
Looks like it.